include/common/intops.rs

```diff
@@ -31,24 +30,39 @@ pub fn umin(a: c_uint, b: c_uint) -> c_uint {
     }
 }
 
+/// # Safety
+///
+/// `U: Into<T>` and `T: TryInto<U>` must be well-formed
+/// such that for all `u: U`, `u.into().try_into() == Ok(u)`.
 #[inline]
-pub fn clip<T: Ord>(v: T, min: T, max: T) -> T {
-    if v < min {
+pub fn clip<T, U>(v: T, min: U, max: U) -> U
+where
+    T: Copy + Ord + TryInto<U>,
+    U: Copy + Ord + Into<T>,
+{
+    assert!(min <= max);
+    if v < min.into() {
         min
-    } else if v > max {
+    } else if v > max.into() {
         max
     } else {
-        v
+        let v = v.try_into();
+        // # Safety
+        // `min <= v <= max`, `min: U`, `max: U`,
+        // and for all `u: U`, `u.into().try_into() == Ok(u)`,
+        // so `v` must be in `U`, too.
+        //
+        // Note that `v.try_into().unwrap()` is not always optimized out.
+        unsafe { v.unwrap_unchecked() }
```

Comment:
Rather than using `try_into` and unwrapping, would it work to use the traits you're adding in #306 to do this as an `as` cast instead? Your logic for why this is safe seems sound to me, but I think we could avoid the unsafety entirely here since this is effectively just doing integer conversions.

Reply:
Hmm, maybe.  I was thinking about that, too, but I was thinking `.as_()` should be restricted to primitives and perhaps we may want a wrapper type for `Pixel` or `Coef` or something.  I'll think about it.

Reply:
I'll come back to this once I've ported more bitdepth-dependent code and have a better idea of what the correct abstraction is, as neither are quite right for this case.